File: lib/workflow_cancel_parallel/workflow_cancel_parallel.ex
defmodule WorkflowCancelParallel do
  @external_resource "docs/workflow_cancel_parallel.md"
  @moduledoc TemporalSdk.Utils.exdoc!("docs/workflow_cancel_parallel.md")

  @spec run(cancellation_delay :: pos_integer()) :: :temporal_sdk.workflow_result() | no_return()
  def run(cancellation_delay) when is_integer(cancellation_delay) do
    {:ok, %{workflow_execution: we}} =
      TemporalSdk.start_workflow(:cluster_1, "default", WorkflowCancelParallel.Workflow)

    Process.sleep(cancellation_delay)
    {:ok, %{}} = TemporalSdk.cancel_workflow(:cluster_1, we)
    {:ok, result} = TemporalSdk.await_workflow(:cluster_1, we)
    result
  end
end

File: src/workflow_cancel_parallel/workflow_cancel_parallel.erl
-module(workflow_cancel_parallel).

% elp:ignore W0012 W0040 E1599
-moduledoc {file, "../../docs/workflow_cancel_parallel.md"}.

-export([
    run/1
]).

-spec run(CancellationDelay :: pos_integer()) -> temporal_sdk:workflow_result() | no_return().
run(CancellationDelay) when is_integer(CancellationDelay) ->
    {ok, #{workflow_execution := WE}} = temporal_sdk:start_workflow(
        cluster_1, "default", workflow_cancel_parallel_workflow
    ),
    timer:sleep(CancellationDelay),
    {ok, #{}} = temporal_sdk:cancel_workflow(cluster_1, WE),
    {ok, Result} = temporal_sdk:await_workflow(cluster_1, WE),
    Result.
